README.md
# NAME

Mojolicious::Plugin::BootstrapPagination - Page Navigator plugin for Mojolicious
This module has derived from [Mojolicious::Plugin::PageNavigator](https://metacpan.org/pod/Mojolicious::Plugin::PageNavigator)

# SYNOPSIS

    # Mojolicious::Lite
    plugin 'bootstrap_pagination'

    # Mojolicious
    $self->plugin( 'bootstrap_pagination' );

# DESCRIPTION

[Mojolicious::Plugin::BootstrapPagination](https://metacpan.org/pod/Mojolicious::Plugin::BootstrapPagination) generates standard page navigation bar, like 

<<  1  2 ... 11 12 13 14 15 ... 85 86 >>

# HELPERS

## bootstrap\_pagination

    %= bootstrap_pagination( $current_page, $total_pages, $opts );

### Options

Options is a optional ref hash.

    %= bootstrap_pagination( $current_page, $total_pages, {
        round => 4,
        outer => 2,
        query => "&id=$id",
        start => 1,
        class => 'pagination-lg',
        param => 'page' } );

- round

    Number of pages around the current page. Default: 4.

- outer

    Number of outer window pages (first and last pages). Default 2.

- param

    Name of param for query url. Default: 'page'

- query

    Additional query string to url. Optional.

- start

    Start number for query string. Default: 1. Optional.

# INTERNATIONALIZATION

If you want to use internationalization (I18N), you can pass a coderef via _localize_.

    plugin 'bootstrap_pagination' => {
      localize => \&localize,
    };
    
    sub localize {
      my ($self, $number) = @_;
    
      my %trans = (
        1 => 'one',
        2 => 'two',
        6 => 'six',
        7 => 'seven',
        8 => 'eight',
        9 => 'nine',
       10 => 'ten',
       11 => 'eleven',
       12 => 'twelve',
       13 => 'thirteen',
       14 => 'fourteen',
       15 => 'fifteen',
      );
    
      return $trans{$number};
    }

This will print the words instead of the numbers.

You can use with Mojolicious::Plugin::I18N.

    plugin 'I18N' => {
      namespace => "Yourpackage::I18N"
    };

    plugin 'bootstrap_pagination' => {
      localize => \&localize,
    };
    
    sub localize {
      my ($self, $number) = @_;
      return $self->l($number);
    }


# SEE ALSO

[Mojolicious](https://metacpan.org/pod/Mojolicious), [Mojolicious::Guides](https://metacpan.org/pod/Mojolicious::Guides), [http://mojolicio.us](http://mojolicio.us),[Mojolicious::Plugin::PageNavigator](https://metacpan.org/pod/Mojolicious::Plugin::PageNavigator).
[Mojolicious::Plugin::I18N](https://metacpan.org/pod/Mojolicious::Plugin::I18N)

# Repository

https://github.com/dokechin/Mojolicious-Plugin-BootstrapPagination

# LICENSE

Copyright (C) dokechin.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

dokechin <>

# CONTRIBUTORS

Andrey Chips Kuzmin <chipsoid@cpan.org>

Renee Baecker <module@renee-baecker.de>





Changes
Revision history for Perl extension Mojolicious-Plugin-BootstrapPagination

{{$NEXT}}

0.13 2015-01-08T12:20:05Z

    - Add I18N feature  
    - allow to set configs when plugin is registered  

0.12 2014-10-25T00:47:43Z

    - Inherit query parameters from the current request   
    - Replace `..` in slice with `…`  c99358b 
    - Changed markup for active page in Bootstrap 3
 
0.11 2014-08-28T09:40:07Z

    - add options

0.10 2014-08-28T23:28:40Z

    - add more options

0.09 2014-03-31T12:20:23Z

    - fix document

0.08 2014-03-12T09:46:56Z

    - add dependency to test

0.07 2014-03-04T23:01:14Z

    - add document

0.06 2014-03-04T11:47:34Z

    - add document

0.05 2014-03-03T23:03:09Z
    
    - add tests

0.04 2014-03-03T10:47:45Z
    - add class option    

0.03 2014-03-02T20:04:42Z

    - change helper name

0.02 2014-03-02T19:54:15Z

    - fix document     

0.01 2014-03-02T19:41:49Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Mojolicious-Plugin-BootstrapPagination',
    module_name     => 'Mojolicious::Plugin::BootstrapPagination',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,


);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

use File::Copy;

print "cp META.json MYMETA.json\n";
copy("META.json","MYMETA.json") or die "Copy failed(META.json): $!";

if (-f 'META.yml') {
    print "cp META.yml MYMETA.yml\n";
    copy("META.yml","MYMETA.yml") or die "Copy failed(META.yml): $!";
} else {
    print "There is no META.yml... You may install this module from the repository...\n";
}



lib/Mojolicious/Plugin/BootstrapPagination.pm
package Mojolicious::Plugin::BootstrapPagination;
use Mojo::Base 'Mojolicious::Plugin';
use POSIX( qw/ceil/ );
use Mojo::ByteStream 'b';

use strict;
use warnings;

our $VERSION = "0.12";

# Homer: Well basically, I just copied the plant we have now.
#        Then, I added some fins to lower wind resistance.  
#        And this racing stripe here I feel is pretty sharp.
# Burns: Agreed.  First prize!
sub  register{
  my ( $self, $app, $args ) = @_;
  $args ||= {};

  $app->helper( bootstrap_pagination => sub{
      my ( $self, $actual, $count, $opts ) = @_;

      my $localize = ( $opts->{localize} || $args->{localize} ) ?
          ( $opts->{localize} || $args->{localize} ) : undef;

      $count = ceil($count);
      return "" unless $count > 1;
      $opts = {} unless $opts;
      my $round = $opts->{round} || $args->{round} || 4;
      my $param = $opts->{param} || $args->{param} || "page";
      my $class = $opts->{class} || $args->{class} || "";
      if ($class ne ""){
          $class = " " . $class;
      }
      my $outer = $opts->{outer} || $args->{outer} || 2;
      my $query = exists $opts->{query} ? $opts->{query} : $args->{query} || "";
      my $start = $opts->{start} // $args->{start} // 1;
      my @current = ( $actual - $round .. $actual + $round );
      my @first   = ($start.. $start + $outer - 1);
      my @tail    = ( $count - $outer + 1 .. $count );
      my @ret = ();
      my $last = undef;
      foreach my $number( sort { $a <=> $b } @current, @first, @tail ){
        next if ( $last && $last == $number && $start > 0 ) || ( defined $last && $last == $number && $start == 0 );
        next if ( $number <= 0 && $start > 0) || ( $number < 0 && $start == 0 );
        last if ( $number > $count && $start > 0 ) || ( $number >= $count && $start == 0 );
        push @ret, ".." if( $last && $last + 1 != $number );
        push @ret, $number;
        $last = $number;
      }
      my $html = "<ul class=\"pagination$class\">";
      if( $actual == $start ){
        $html .= "<li class=\"disabled\"><a href=\"#\" >&laquo;</a></li>";
      } else {
        $html .= "<li><a href=\"" . $self->url_with->query( [$param => $actual - 1] ) . $query . "\" >&laquo;</a></li>";
      }
      my $last_num = -1;
      foreach my $number( @ret ){
        my $show_number = $start > 0 ? $number : ( $number =~ /\d+/ ? $number + 1 : $number );

        if ( $localize ) {
            $show_number = $localize->($self, $show_number);
        }

        if( $number eq ".." && $last_num < $actual ){
          my $offset = ceil( ( $actual - $round ) / 2 ) + 1 ;
          $html .= "<li><a href=\"" . $self->url_with->query( [$param => $start == 0 ? $offset + 1 : $offset] ) . $query ."\" >&hellip;</a></li>";
        }
        elsif( $number eq ".." && $last_num > $actual ) {
          my $back = $count - $outer + 1;
          my $forw = $round + $actual;
          my $offset = ceil( ( ( $back - $forw ) / 2 ) + $forw );
          $html .= "<li><a href=\"" . $self->url_with->query( [$param => $start == 0 ? $offset + 1 : $offset] ) . $query ."\" >&hellip;</a></li>";
        } elsif( $number == $actual ) {
          $html .= "<li class=\"active\"><span>$show_number</span></li>";
        } else {
          $html .= "<li><a href=\"" . $self->url_with->query( [$param => $number] ) . $query ."\">$show_number</a></li>";
        }
         $last_num = $number;
      }
      if( $actual == $count ){
        $html .= "<li class=\"disabled\"><a href=\"" . $self->url_with->query( [$param => $actual + 1] ) . $query . "\" >&raquo;</a></li>";
      } else {
        $html .= "<li><a href=\"" . $self->url_with->query( [$param => $actual + 1] ) . $query . "\" >&raquo;</a></li>";
      }
      $html .= "</ul>";
      return b( $html );
    } );

}

1;
__END__

=encoding utf-8

=head1 NAME

Mojolicious::Plugin::BootstrapPagination - Page Navigator plugin for Mojolicious
This module has derived from L<Mojolicious::Plugin::PageNavigator>

=head1 SYNOPSIS

  # Mojolicious::Lite
  plugin 'bootstrap_pagination'

  # Mojolicious
  $self->plugin( 'bootstrap_pagination' );

=head1 DESCRIPTION

L<Mojolicious::Plugin::BootstrapPagination> generates standard page navigation bar, like 
  
<<  1  2 ... 11 12 13 14 15 ... 85 86 >>

=head1 HELPERS

=head2 bootstrap_pagination

  %= bootstrap_pagination( $current_page, $total_pages, $opts );

=head3 Options

Options is a optional ref hash.

  %= bootstrap_pagination( $current_page, $total_pages, {
      round => 4,
      outer => 2,
      query => "&id=$id",
      start => 1,
      class => 'pagination-lg',
      param => 'page' } );

=over 1

=item round

Number of pages around the current page. Default: 4.

=item outer

Number of outer window pages (first and last pages). Default 2.

=item param

Name of param for query url. Default: 'page'

=item query

Additional query string to url. Optional.

=item start

Start number for query string. Default: 1. Optional.

=back

=head1 INTERNATIONALIZATION

If you want to use internationalization (I18N), you can pass a coderef via I<localize>.

  plugin 'bootstrap_pagination' => {
    localize => \&localize,
  };
  
  sub localize {
    my ($number) = @_;
  
    my %trans = (
      1 => 'one',
      2 => 'two',
      6 => 'six',
      7 => 'seven',
      8 => 'eight',
      9 => 'nine',
     10 => 'ten',
     11 => 'eleven',
     12 => 'twelve',
     13 => 'thirteen',
     14 => 'fourteen',
     15 => 'fifteen',
    );
  
    return $trans{$number};
  }

This will print the words instead of the numbers.

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Guides>, L<http://mojolicio.us>,L<Mojolicious::Plugin::PageNavigator>.

=head1 Repository

https://github.com/dokechin/Mojolicious-Plugin-BootstrapPagination

=head1 LICENSE

Copyright (C) dokechin.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

dokechin E<lt>E<gt>

=head1 CONTRIBUTORS

Andrey Chips Kuzmin <chipsoid@cpan.org>

=cut

